renpy/curry.py

```diff
@@ -23,54 +23,68 @@
 
+import functools
 
 
 class Curry(object):
     """
-    Stores a callable and some arguments. When called, calls the
-    callable with the stored arguments and the additional arguments
-    supplied to the call.
+    :undocumented:
+
+    Compatibilty class that exists solely to deserialise instances from
+    old saves into instances of the replacement Partial type below.
     """
 
-    hash = None
+    def __new__(cls, *args, **kwargs):
+        if args:
+            return Partial(*args, **kwargs)
 
-    def __init__(self, callable, *args, **kwargs):  # @ReservedAssignment
-        self.callable = callable
-        self.args = args
-        self.kwargs = kwargs
-        self.__doc__ = getattr(self.callable, "__doc__", None)
-        self.__name__ = getattr(self.callable, "__name__", None)
+        # No args supplied, so likely we're trying to unpickle an instance
+        # from an old save. Create a new Partial with an arbitrary function
+        # which will be updated in a moment by __setstate__.
+        rv = Partial(hash)
+        rv.__setstate__ = cls.__setstate__.__get__(rv)
 
-    def __call__(self, *args, **kwargs):
+        return rv
 
-        merged_kwargs = dict(self.kwargs)
-        merged_kwargs.update(kwargs)
+    def __setstate__(self, state):
+        state.pop('hash', None)
+        state = (state.pop('callable'), state.pop('args'), state.pop('kwargs'), state)
+        return super(Partial, self).__setstate__(state)
 
-        return self.callable(*(self.args + args), **merged_kwargs)
+
+class Partial(functools.partial):
+    """
+    Stores a callable and some arguments. When called, calls the
+    callable with the stored arguments and the additional arguments
+    supplied to the call.
+    """
+
+    __slots__ = ('hash',)
 
     def __repr__(self):
-        return "<curry %s %r %r>" % (self.callable, self.args, self.kwargs)
+        return "<partial %s %r %r>" % (self.func, self.args, self.keywords)
 
     def __eq__(self, other):
-
-        return (
-            isinstance(other, Curry) and
-            self.callable == other.callable and
-            self.args == other.args and
-            self.kwargs == other.kwargs)
+        return (isinstance(other, Partial) and
+                self.func == other.func and
+                self.args == other.args and
+                self.keywords == other.keywords)
 
     def __ne__(self, other):
         return not (self == other)
 
     def __hash__(self):
+        _hash = getattr(self, 'hash', None)
+
+        if _hash is None:
+            _hash = hash(self.func) ^ hash(self.args)
 
-        if self.hash is None:
-            self.hash = hash(self.callable) ^ hash(self.args)
+            for i in self.keywords.items():
+                _hash ^= hash(i)
 
-            for i in self.kwargs.items():
-                self.hash ^= hash(i)
+            setattr(self, 'hash', _hash)
```

Comment:
```suggestion
            self.hash = _hash
```

Reply:
Wrote it this way to have it mirror the `getattr` used the start of the function which I find helps in recognising that the same thing being accessed/written. Also ends up having (nominally) better performance.